rfcs/0053-reducing-daemon-memory-usage.md

```diff
@@ -0,0 +1,137 @@
+
+Now, if we adjust the estimates for proposed optimizations to store proofs and verification keys to disk (by subtracting their sizes from the memory footprint and replacing them with cache references), we get the following output.
+
+```
+baseline = 3.064569GB
+scan_states = 3.658435GB
+ledger_masks = 0.562126GB
+staged_ledger_diffs = 9.202166GB
+snark_pool = 0.014414GB
+transaction_pool = 0.247903GB
+TOTAL: 16.749612GB
+```
+
+As we can see, this brings the estimation down much closer to the current `16GB` hardware requirement. From here, we can look into additional optimizations to bring it down even further to fit within the current hardware requirements. Such optimizations could include:
+
+* sharing the prover key allocation across the daemon's subprocesses, reducing the baseline to nearly 1/3rd of what it is now
+* persisting the entire staged ledger diff of each block to disk, given we rarely need to read the commands contained within a staged ledger diff after the diff has been applied
+* throwing out older proofs/commands from scan states once they are no longer required for the remaining work to be included
+
+### Impact Analysis
+[impact-analysis]: #impact-analysis
+
+There is a space-time tradeoff here in the sense that, by excising data from RAM to disk, we are now needing to perform disk I/O and deserialize/serialize data in order to perform reads/writes. So part as part of this design, it is important to show that the performance hit we take for reading/writing data cached to disk is relatively insignificant in the operation of the daemon.
+
+<!--
+I tried to use the lmdb benchmarks to get accurate numbers for the exact sizes of our data structures, but was unable to get the program working with large sizes for some reason. I can spend some time later to make the benchmark program work, if necessary.
+-->
+
+LMDB provides benchmarks against other similar databases [on their website](http://www.lmdb.tech/bench/microbench/). The important piece of data here is that, with a 128MB cache and values of `100,000` bytes in size, LMDB is benched at being capable of performing `1,718,213` random reads per second (about `582` nanoseconds per read). Given the amount of reads, and frequencey of reads, a daemon will be performing from this on-disk cache, these benchmarks show that reading from LMDB will have a negligible effect on daemon performance. All proofs and verification keys we would read/write to disk are under this `100,000` benchmark size, so the actual performance should be better than this.
```

Comment:
I expect the latency to be more relevant here than the throughput- there's basically a floor on performance due to transmission latencies between the platter and the CPU cache, and that very well might dominate the cost of small reads/writes. We will be completely at the mercy of the operating system's page cache when it comes down to eviction/refilling time- the below seems to assume a given datum need be read from disk only once, but in fact it may occur many times during high memory pressure.

That said I don't think a more detailed analysis will change anyone's mind that we want to do this, this is good evidence the impact will be acceptable.

Do we have good enough latency instrumentation that we think we could notice the end-to-end impact with a simple statistical test after it is implemented?

Reply:
I was thinking that this benchmark should show the average case latency since it is using random accesses, so it should show an even load of cache misses and page loads, but maybe that's a bad assumption. I won't do any more detailed analysis for now, but let me know if you believe it is warranted.

> Do we have good enough latency instrumentation that we think we could notice the end-to-end impact with a simple statistical test after it is implemented?

We can utilize Viable System's latency metrics from openmina to detect a performance regression.